ENH: generalize hough_peak functions

sciunto opened the pull request:
Fixes #1964

CHANGELOG:
- [x] `hough_line_peaks()` becomes `_prominent_peaks()`, a general function to find peaks that we can improve in the future.
- [x] `hough_line_peaks()` is now a wrapper of `prominent_peaks()`
- [x] test for `prominent_peaks()` added
- [x] new function `hough_circle_peaks()` that reuse `prominent_peaks()`
- [x] test `hough_circle_peaks()`
- [x] update gallery

ahojnnes commented on skimage/feature/peak.py:
```diff
@@ -173,3 +174,106 @@ def peak_local_max(image, min_distance=1, threshold_abs=None,
         return out
+
+
+def prominent_peaks(image, min_xdistance=1, min_ydistance=1,
```
This should be generalizable to N-D easily, I assume?

sciunto replied:
I'm not familiar with that. Do we have a doc? Didn't see anything in the user guide...

sciunto commented:
@emmanuelle OK, this is the PR I was mentionning to you. Actually, the difference is about the flatness of peaks as explained in the docstring.

sciunto commented:
Looking at it again, I think I'll turn `prominent_peak` into a private function. I think it disserves more work, but it goes beyond this PR.

sciunto commented:
@scikit-image/core I rebased my PR. Anyone to review? this one is getting old. :)

sciunto commented:
I close this PR because I don't want to work on this anymore. It's open for more than 4 months, no feedback and I don't want to rebase over and over. If anyone is interested, he/she can reuse it.

sciunto commented:
I ping again @scikit-image/core to check if anyone changed is mind. I still think this represents a major enhancement...

blink1073 commented:
@sciunto, can you add a screen shot using the new `hough_circle_peaks`?

sciunto commented:
There it is
![1](https://cloud.githubusercontent.com/assets/335370/22758062/54a4d396-ee4d-11e6-827f-5cdc21fe208c.png)

I also rebased and fixed some stuffs.

blink1073 commented:
Cool, looks good, thanks.

soupault commented on skimage/feature/peak.py:
```diff
@@ -1,22 +1,24 @@
-    return np.transpose(coord)[sorted_indices]
+        intensities = image[coord]
+        idx_maxsort = np.argsort(intensities)
+        coord = np.transpose(coord)[idx_maxsort][-num_peaks:]
```
I'm not a user of this function, but does the order of coords matter here (was num_peaks sorted from most to least intensity, now reversed)?

sciunto replied:
I reverted this. Now, it's highest first. I don't remember why I did that.

soupault commented on skimage/feature/peak.py:
```diff
@@ -183,3 +192,106 @@ def peak_local_max(image, min_distance=1, threshold_abs=None,
+    ycoords_size = 2 * min_ydistance + 1
+    xcoords_size = 2 * min_xdistance + 1
+    img_max = ndi.maximum_filter1d(img, size=ycoords_size, axis=0,
+                                      mode='constant', cval=0)
```
Please, fix indentation here and above.

soupault commented on skimage/feature/tests/test_peak.py:
```diff
@@ -1,426 +1,450 @@
+                                     footprint=np.ones((3, 3), bool),
+                                     min_distance=1, threshold_rel=0,
+                                     indices=False, exclude_border=False)
+        assert np.all(labels == labelsin)
```
This will probably set up a nightmare for Nelle :fearful: . :)

soupault commented on skimage/transform/hough_transform.py:
```diff
@@ -209,26 +160,28 @@ def hough_ellipse(img, threshold=4, accuracy=1, min_size=4, max_size=None):
+                          min_size=min_size, max_size=max_size)
 
 
+# Wrapper for Cython allows function signature introspection
```
I'd remove this comment.

soupault commented on skimage/transform/hough_transform.py:
```diff
@@ -209,26 +160,28 @@ def hough_ellipse(img, threshold=4, accuracy=1, min_size=4, max_size=None):
     img : (M, N) ndarray
         Input image with nonzero values representing edges.
-    theta : 1D ndarray of double, optional
+    theta : 1D ndarray of double
```
Is `optional` still?

soupault commented on skimage/transform/hough_transform.py:
```diff
@@ -258,14 +213,19 @@ def hough_line(img, theta=None):
 
-def probabilistic_hough_line(img, threshold=10, line_length=50,
-                             line_gap=10, theta=None):
+# Wrapper for Cython allows function signature introspection
```
Remove also :). I mean, we have a large number of wrappers, no sense to me to highlight this for all of them.

soupault commented on skimage/transform/hough_transform.py:
```diff
@@ -287,8 +247,8 @@ def probabilistic_hough_line(img, threshold=10, line_length=50,
     lines : list
-        List of lines identified, lines in format ((x0, y0), (x1, y1)),
-        indicating line start and end.
+      List of lines identified, lines in format ((x0, y0), (x1, y0)),
```
This was correctly fixed earlier, should revert.

soupault commented on skimage/transform/hough_transform.py:
```diff
@@ -301,7 +261,90 @@ def probabilistic_hough_line(img, threshold=10, line_length=50,
+    """Return peaks in a circle Hough transform.
+
+    Identifies most prominent circles separated by certain distances
+    in a Hough transform. Non-maximum suppression with different sizes is
```
`in a Hough transform.` -> `in a Hough space.`?

soupault commented on skimage/transform/hough_transform.py:
```diff
@@ -301,7 +261,90 @@ def probabilistic_hough_line(img, threshold=10, line_length=50,
+        Hough spaces returned by the `hough_circle` function.
+    radii : (M,) array
+        Radii corresponding to Hough spaces.
+    min_xdistance : int
```
Would be good to add `, optional` for all arguments.

sciunto commented:
Thanks! I'll finish this over the week-end. If it could go asap in master, that would be awesome.

soupault commented:
Thanks @sciunto !

blink1073 commented:
:tada:
